test/chol.jl

```diff
@@ -1,49 +1,50 @@
-using StaticArrays, Compat, Compat.Test, Compat.LinearAlgebra
+using StaticArrays, Test, LinearAlgebra
+using LinearAlgebra: PosDefException
 
 @testset "Cholesky decomposition" begin
-    if !isdefined(Compat.LinearAlgebra, :non_hermitian_error)
-        PosDefException = Compat.LinearAlgebra.PosDefException
-    else
-        PosDefException = ArgumentError
-    end
     @testset "1×1" begin
         m = @SMatrix [4.0]
-        (c,) = chol(m)
+        (c,) = cholesky(m).U
         @test c === 2.0
     end
 
     @testset "2×2" for i = 1:100
         m_a = randn(2,2)
         #non hermitian
-        @test_throws PosDefException chol(SMatrix{2,2}(m_a))
+        @test_throws PosDefException cholesky(SMatrix{2,2}(m_a))
         m_a = m_a*m_a'
         m = SMatrix{2,2}(m_a)
-        @test chol(Hermitian(m)) ≈ chol(m_a)
+        @test cholesky(Hermitian(m)).U ≈ cholesky(m_a).U
+        @test cholesky(Hermitian(m)).L ≈ cholesky(m_a).L
     end
 
     @testset "3×3" for i = 1:100
         m_a = randn(3,3)
         #non hermitian
-        @test_throws PosDefException chol(SMatrix{3,3}(m_a))
+        @test_throws PosDefException cholesky(SMatrix{3,3}(m_a))
         m_a = m_a*m_a'
         m = SMatrix{3,3}(m_a)
-        @test chol(m) ≈ chol(m_a)
-        @test chol(Hermitian(m)) ≈ chol(m_a)
+        @test cholesky(m).U ≈ cholesky(m_a).U
+        @test cholesky(m).L ≈ cholesky(m_a).L
+        @test cholesky(Hermitian(m)).U ≈ cholesky(m_a).U
+        @test cholesky(Hermitian(m)).L ≈ cholesky(m_a).L
     end
     @testset "4×4" for i = 1:100
         m_a = randn(4,4)
         #non hermitian
-        @test_throws PosDefException chol(SMatrix{4,4}(m_a))
+        @test_throws PosDefException cholesky(SMatrix{4,4}(m_a))
         m_a = m_a*m_a'
         m = SMatrix{4,4}(m_a)
-        @test chol(m) ≈ chol(m_a)
-        @test chol(Hermitian(m)) ≈ chol(m_a)
+        @test cholesky(m).L ≈ cholesky(m_a).L
+        @test cholesky(m).U ≈ cholesky(m_a).U
+        @test cholesky(Hermitian(m)).L ≈ cholesky(m_a).L
+        @test cholesky(Hermitian(m)).U ≈ cholesky(m_a).U
     end
     @testset "static blockmatrix" for i = 1:10
         m_a = randn(3,3)
         m_a = m_a*m_a'
         m = SMatrix{3,3}(m_a)
-        @test chol(reshape([m, 0m, 0m, m], 2, 2)) ==
+        @test_broken cholesky(reshape([m, 0m, 0m, m], 2, 2)) ==
```

Comment:
This `cholesky` call causes a call to `hermitian_type(::Type{Any}) which fails. @martinholters you implemented that, any idea why this happens?

Reply:
`cholesky` calls `cholcopy`, which gives a `Matrix{Any}`. Probably `choltype`, responsible for figuring out the element type, should be made recursive. Note that `cholesky(reshape([m_a, 0m_a, 0m_a, m_a], 2, 2))` also fails for the same underlying reason, except that here, `choltype` fails itself. Probably worth an issue in Julia.

Reply:
Let's deal with the universe of block matrices afterwards.